File: C++/solving-questions-with-brainpower.cpp
// Time:  O(n)
// Space: O(n)

// dp
class Solution {
public:
    long long mostPoints(vector<vector<int>>& questions) {
        vector<int64_t> dp(size(questions) + 1);
        for (int i = size(dp) - 2; i >= 0; --i) {
            dp[i] = max(dp[i + 1], questions[i][0] + (i + 1 + questions[i][1] < size(dp) ? dp[i + 1 + questions[i][1]] : static_cast<int64_t>(0)));
        }
        return dp[0];
    }
};

File: Python/solving-questions-with-brainpower.py
# Time:  O(n)
# Space: O(n)

# dp
class Solution(object):
    def mostPoints(self, questions):
        """
        :type questions: List[List[int]]
        :rtype: int
        """
        dp = [0]*(len(questions)+1)
        for i in reversed(xrange(len(dp)-1)):
            dp[i] = max(dp[i+1], questions[i][0] + (dp[i+1+questions[i][1]] if i+1+questions[i][1] < len(dp) else 0))
        return dp[0]
